### ISSUE TITLE
Bug in get_next_power_of_2

### ISSUE THREAD

**Issue by franck44** [ethereum/eth2.0-specs#1695]
The [simple-serialise](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/simple-serialize.md) (Merkleisation section) reads:

> next_pow_of_two(i): get the next power of 2 of i, if not already a power of 2, with 0 mapping to 1. Examples: 0->1, 1->1, 2->2, 3->4, 4->4, 6->8, 9->16

In the [merkle proofs](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/merkle-proofs.md), helpers are defined by:

```
def get_next_power_of_two(x: int) -> int:
    """
    Get next power of 2 >= the input.
    """
    if x <= 2:
        return x
    else:
        return 2 * get_next_power_of_two((x + 1) // 2)

def get_previous_power_of_two(x: int) -> int:
    """
    Get the previous power of 2 >= the input.
    """
    if x <= 2:
        return x
    else:
        return 2 * get_previous_power_of_two(x // 2)
```
Unfortunately, the first one is buggy (e.g. for 0 it returns 0 which is not a power of 2 and is different to what the simple-serialise specs reads.)
Here is a fix, proved correct with Dafny (the fact that it computes the next power of 2 of a given natural number requires more lemmas, not included in this post):
```
function get_next_power_of_two(n : int) : int 
    requires n >= 0
    ensures get_next_power_of_two(n) >= 1
{
    if n <= 1 then 1
    else 2 * get_next_power_of_two( (n + 1) / 2)
}
```

The second one ` get_previous_power_of_two` is not clearly defined. 
Anyway it seems buggy as well as it returns 0 for 0 which is not a power of 2.
What is the previous power of 2 >= input? Is it the power of 2, say k, such that 2^k < input <= 2^{k+1}?
Is it the largest power of 2 that is strictly smaller than the input? If yes, we also have a (provably) correct version of it.

**Comment by protolambda** [comment]
The naming is pretty terrible here (different names for two very similar functions). And then for the incorrect one: in one place (`get_generalized_index`) it indeed should return 1 (but it will never be called with 0, so it does not matter), and in the other (`merkle_tree`) it should be 0 (no extra padding necessary, although a tree with 0 leaves is quite strange, maybe padding of 1 is fine as default there). And then it does not seem to be used anywhere else in the specs.

I'm not opposed to changing it to handle the `0 -> 1` properly, and unifying the names. That said, the `ssz` folder is planned to move out, to be replaced with that ssz draft repo once that is updated, and then we won't have this buggy function anymore.

Thanks for documenting this, I think I'll prioritize bringing the draft up to date and online so we won't have these issues anymore. :+1:

**Comment by franck44** [comment]
Thanks for that. 
It is probably safer to document it and even it is not called with 0 now, it may happen in the future.
But overall, what is the mathematical spec of these functions?

This is what we have in our Dafny code at the moment:
```
 /** Get the next power of two.
     *
     *  @param  n   A positive integer. 
     *  @return     The smallest power of 2 that is larger than n.
     */
    function get_next_power_of_two(n : nat) : nat 
    requires true // n is a nat which is an int >=0 so pre-condition is true.
    ensures get_next_power_of_two(n) >= 1
    ensures get_next_power_of_two(n) >= n + 1 // result should be > n
    ensures n >= 1 ==> (get_next_power_of_two(n) / 2) < n // previous power of two should be < n
```

**Comment by franck44** [comment]
@protolambda We have opened our repository on formal verification of Eth2.0 using Dafny.
Here are [power_of_2 spec and proofs](https://github.com/PegaSysEng/eth2.0-dafny/blob/d3a3e8b799e2a4b8bac12ec5ccc9ef6123842f53/src/dafny/utils/MathHelpers.dfy)

**Comment by protolambda** [comment]
See #1907, the code is not used anywhere really, but I made an illustrative fix, while we figure out #1901.
